File: arrays_hashing/two_sum/two_sum_test.go
package twosum

import (
	"slices"
	"testing"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		name     string
		nums     []int
		target   int
		expected []int
	}{
		{
			name:     "Example 1: Basic case",
			nums:     []int{3, 4, 5, 6},
			target:   7,
			expected: []int{0, 1},
		},
		{
			name:     "Example 2: Gap in indices",
			nums:     []int{4, 5, 6},
			target:   10,
			expected: []int{0, 2},
		},
		{
			name:     "Example 3: Duplicate numbers",
			nums:     []int{5, 5},
			target:   10,
			expected: []int{0, 1},
		},
		{
			name:     "Large numbers",
			nums:     []int{1000000, 500, 2000000},
			target:   3000000,
			expected: []int{0, 2},
		},
		{
			name:     "Target is first two elements",
			nums:     []int{1, 2, 3, 4, 5},
			target:   3,
			expected: []int{0, 1},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := TwoSum(tt.nums, tt.target)
			if !slices.Equal(got, tt.expected) {
				t.Errorf("TwoSum() = %v, want %v", got, tt.expected)
			}
		})
	}
}

File: arrays_hashing/two_sum/two_sum_test.py
import pytest
from two_sum import two_sum


@pytest.mark.parametrize(
    "nums, target, expected",
    [
        ([3, 4, 5, 6], 7, [0, 1]),  # Example 1
        ([4, 5, 6], 10, [0, 2]),  # Example 2
        ([5, 5], 10, [0, 1]),  # Example 3: Duplicates
        ([1, 2, 3, 4], 3, [0, 1]),  # Target at start
        ([1, 3, 4, 2], 6, [2, 3]),  # Target at end
        ([1000, 2000], 3000, [0, 1]),  # Large numbers
    ],
)
def test_two_sum(nums, target, expected):
    """
    Ensures two_sum returns the correct pair of indices
    in the required order (smaller index first).
    """
    result = two_sum(nums, target)
    assert result == expected
